src/sentry/integrations/models/external_issue.py

```diff
@@ -0,0 +1,98 @@
+from sentry.eventstore.models import Event
+
+if TYPE_CHECKING:
+    from django.db.models.query import _QuerySet
+
+    from sentry.integrations.services.integration import RpcIntegration
+
+
+class ExternalIssueManager(BaseManager["ExternalIssue"]):
+    def get_for_integration(
+        self, integration: RpcIntegration, external_issue_key: str | None = None
+    ) -> QuerySet:
+        from sentry.integrations.services.integration import integration_service
+
+        org_integrations = integration_service.get_organization_integrations(
+            integration_id=integration.id
+        )
+
+        kwargs = dict(
+            integration_id=integration.id,
+            organization_id__in=[oi.organization_id for oi in org_integrations],
+        )
+
+        if external_issue_key is not None:
+            kwargs["key"] = external_issue_key
+
+        return self.filter(**kwargs)
+
+    def get_linked_issues(
+        self, event: Event, integration: RpcIntegration
+    ) -> QuerySet[ExternalIssue]:
+        from sentry.models.grouplink import GroupLink
+
+        assert event.group is not None
+        return self.filter(
+            id__in=GroupLink.objects.filter(
+                project_id=event.group.project_id,
+                group_id=event.group.id,
+                linked_type=GroupLink.LinkedType.issue,
+            ).values_list("linked_id", flat=True),
+            integration_id=integration.id,
+        )
+
+    def get_linked_issue_ids(
+        self, event: Event, integration: RpcIntegration
+    ) -> _QuerySet[ExternalIssue, str]:
+        return self.get_linked_issues(event, integration).values_list("key", flat=True)
+
+    def has_linked_issue(self, event: Event, integration: RpcIntegration) -> bool:
+        return self.get_linked_issues(event, integration).exists()
+
+
+@region_silo_model
+class ExternalIssue(Model):
+    __relocation_scope__ = RelocationScope.Excluded
+
+    # The foreign key here is an `int`, not `bigint`.
+    organization = FlexibleForeignKey("sentry.Organization", db_constraint=False)
+
+    integration_id = HybridCloudForeignKey("sentry.Integration", on_delete="CASCADE")
+
+    key = models.CharField(max_length=256)  # example APP-123 in jira
+    date_added = models.DateTimeField(default=timezone.now)
+    title = models.TextField(null=True)
+    description = models.TextField(null=True)
+    metadata = JSONField(null=True)
+
+    objects: ClassVar[ExternalIssueManager] = ExternalIssueManager()
+
+    class Meta:
+        app_label = "sentry"
+        db_table = "sentry_externalissue"
+        unique_together = (("organization", "integration_id", "key"),)
+
+    __repr__ = sane_repr("organization_id", "integration_id", "key")
+
+    def get_installation(self) -> Any:
+        from sentry.integrations.services.integration import integration_service
+
+        integration = integration_service.get_integration(integration_id=self.integration_id)
+
+        return (
```

Comment:
I changed this btw from `return integration.get_installation(organization_id=self.organization_id)` to deal with mypy saying `None doesn't have .get_integration()`

Reply:
Makes sense. Another solution for these kinds of error where the types indicate that something could be missing, but it rarely is in practice is to use assertions

```python
assert integration, "integration is required to get an installation"
```

an assertion like this will narrow the variable's type removing `None`. Raising an exception can be helpful in identifiying if this problem occurs in production more than returning a `None` will.